src/os/bluestore/BlueStore.cc

```diff
@@ -7139,6 +7139,9 @@ void BlueStore::_txc_state_proc(TransContext *txc)
 	} else {
 	  _txc_finalize_kv(txc, txc->t);
 	  txc->state = TransContext::STATE_KV_SUBMITTED;
+	  if (txc->osr->kv_submitted_waiters) {
```

Comment:
Originally you did that notification a bit better - all txc-s were enumerated first and then notification sent. The only thing it lacked - more precise detection if notify to be sent. But right now you might do multiple notifications that are no-op until the last entry in the queue is 'kv_submitted'. Hence the same ineffectiveness as I pointed out in the below comment...